misc/python/materialize/data_ingest/definition.py

```diff
@@ -0,0 +1,185 @@
+from materialize.data_ingest.field import Field
+from materialize.data_ingest.row import Operation, Row
+from materialize.data_ingest.rowlist import RowList
+from materialize.data_ingest.transaction import Transaction
+
+
+class Records(Enum):
+    ONE = 1
+    MANY = 2
+    ALL = 3
+
+
+class Keyspace(Enum):
+    SINGLE_VALUE = 1
+    LARGE = 2
+    EXISTING = 3
+
+
+class Target(Enum):
+    KAFKA = 1
+    POSTGRES = 2
+    PRINT = 3
+
+
+class Definition:
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        raise NotImplementedError
+
+
+class Insert(Definition):
+    def __init__(self, count: Records, record_size: RecordSize):
+        self.count = count
+        self.record_size = record_size
+        self.current_key = 0
+
+    def max_key(self) -> int:
+        if self.count == Records.ONE:
+            return 1
+        elif self.count == Records.MANY:
+            return 1000
+        else:
+            raise ValueError(f"Unexpected count {self.count}")
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        if self.count == Records.ONE:
+            count = 1
+        elif self.count == Records.MANY:
+            count = 1000
+        else:
+            raise ValueError(f"Unexpected count {self.count}")
+
+        fields_with_values = deepcopy(fields)
+
+        transactions = []
+        for i in range(count):
+            for field in fields_with_values:
+                if field.is_key:
+                    field.value = field.typ.num_value(self.current_key)
+                else:
+                    field.value = field.typ.random_value(self.record_size)
+            self.current_key += 1
+            transactions.append(
+                Transaction(
+                    [
+                        RowList(
+                            [
+                                Row(
+                                    fields=deepcopy(fields_with_values),
+                                    operation=Operation.INSERT,
+                                )
+                            ]
+                        )
+                    ]
+                )
+            )
+
+        return transactions
+
+
+class Upsert(Definition):
+    def __init__(self, keyspace: Keyspace, count: Records, record_size: RecordSize):
+        self.keyspace = keyspace
+        self.count = count
+        self.record_size = record_size
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        if self.count == Records.ONE:
+            count = 1
+        elif self.count == Records.MANY:
+            count = 1000
+        else:
+            raise ValueError(f"Unexpected count {self.count}")
+
+        fields_with_values = deepcopy(fields)
+
+        transactions = []
+        for i in range(count):
+            for field in fields_with_values:
+                if field.is_key:
+                    if self.keyspace == Keyspace.SINGLE_VALUE:
+                        field.value = field.typ.num_value(0)
```

Comment:
Is it intended to always use `0` as key here? Or should `i` be used?

Reply:
Intentional, we are actually updating the same key again and again.